[dataclasses plugin] Support kw_only=True

tgallant opened the pull request:
### Description

<!--
If this pull request closes or fixes an issue, write Closes #NNN" or "Fixes #NNN" in that exact
format.
-->

Fixes #10865

If all fields in a dataclass are keyword only then mypy should not raise errors regarding the order of attributes with default values. See the linked issue for a more detailed description.

This PR makes the relevant changes to handle that.

## Test Plan

<!--
If this is a documentation change, rebuild the docs (link to instructions) and review the changed pages for markup errors.
If this is a code change, include new tests (link to the testing docs). Be sure to run the tests locally and fix any errors before submitting the PR (more instructions).
If this change cannot be tested by the CI, please explain how to verify it manually.
-->

I've added tests to `check-dataclasses.test` to cover the new changes.

tgallant commented on test-data/unit/lib-stub/dataclasses.pyi:
```diff
@@ -1,30 +1,86 @@
 
+if sys.version_info >= (3, 10):
+    class _KW_ONLY_TYPE: ...
+    KW_ONLY = _KW_ONLY_TYPE
```
in the stub from typeshed this line is

```python
    KW_ONLY: _KW_ONLY_TYPE
```

Without making this change the following will fail

```python
[case testDataclassesOrderingKwOnlyWithSentinel]
# flags: --python-version 3.10
from dataclasses import dataclass, KW_ONLY

@dataclass
class Application:
    _: KW_ONLY
    name: str = 'Unnamed'
    rating: int
```

```
error: Variable "dataclasses.KW_ONLY" is not valid as a type
```

Any feedback on this? Does this change need to be made in typeshed or am I specifying the type wrong in the test case?

JelleZijlstra replied:
I think typeshed has to change. I believe that test case is how the object is supposed to be used, so typeshed should accommodate it.

cdce8p commented:
Came across an interesting use case of `KW_ONLY` with inheritance, that should probably be supported as well (if not already done).
```py
from dataclasses import dataclass, KW_ONLY

@dataclass
class Base:
    x: str
    _: KW_ONLY
    y: int = 0
    w: int = 1

@dataclass
class D(Base):
    z: str
    a: str = "a"
```
```py
>>> D("Hello", "World")
D(x='Hello', y=0, w=1, z='World', a='a')
```
That works because of keyword reordering.
https://docs.python.org/3.10/library/dataclasses.html#re-ordering-of-keyword-only-parameters-in-init

hauntsaninja commented:
Thanks for working on this! Could we also test that `__init__` has the appropriate signature (e.g. by instantiating the classes)

tgallant commented:
Thanks for raising this case @cdce8p. I've added this example as a test case and updated the code to handle it. @hauntsaninja I also made sure to check the initialization in the test cases.

I was looking through the docs again and spotted this:

> In a single dataclass, it is an error to specify more than one field whose type is KW_ONLY.

https://docs.python.org/3.10/library/dataclasses.html#dataclasses.KW_ONLY

It seemed like this would be a good thing to have a mypy check for. I added a check for this case in commit https://github.com/python/mypy/pull/10867/commits/75ef138142b6a7da173730afa28ab49f2e4e2374

cdce8p commented on test-data/unit/lib-stub/dataclasses.pyi:
```diff
@@ -5,26 +5,29 @@ _T = TypeVar('_T')
 class InitVar(Generic[_T]):
     ...
 
+class _KW_ONLY_TYPE: ...
+KW_ONLY = _KW_ONLY_TYPE
```
Not entirely sure, but https://github.com/python/typeshed/pull/5826 changed how `KW_ONLY` is defined.
After the typeshed changes have been synced, I think this needs to be updated.

hauntsaninja commented:
python/typeshed#5826 is now on mypy master, so if you merge you'll be able to pull that in.

It would be good to add tests of class instantiation that produce errors. Specifically, test cases should prove that the arguments are in fact keyword only and complain when passed positionally and that parameter reordering happens and we get type errors (i.e. make sure positional parameters have types that are distinct from each other and from kw only parameters)

tgallant commented:
i rebased this branch with master and updated the `KW_ONLY` usage according to the typeshed change. i will follow up with some more test cases

tgallant commented:
thanks for the helpful suggestions @hauntsaninja and @cdce8p. i've added test cases covering positional vs keyword arguments, reordering, and type errors. let me know if you think there are other cases that should be covered in this PR.

hauntsaninja reviewed (approved):
Thanks, this is great! :-)

cdce8p commented:
Thanks @tgallant 🎉
